File: SJSegmentedScrollView-ObjC/SJContentView.m
//
//  SJContentView.m
//  Hike
//
//  Created by Pavan Goyal on 02/05/18.
//  Copyright © 2018 Hike Pvt. Ltd. All rights reserved.
//

#import "SJContentView.h"

@interface SJContentView () <UIScrollViewDelegate>

@property (nonatomic, assign) NSInteger pageIndex;
@property (nonatomic, strong) NSMutableArray<UIView *> *contentViews;
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) NSLayoutConstraint *contentViewWidthConstraint;
@property (nonatomic, strong) NSMutableArray<NSLayoutConstraint *> *contentSubViewWidthConstraints;
@property (nonatomic, assign) CGFloat animationDuration;

@end

@implementation SJContentView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _pageIndex = 0;
        _contentViews = [[NSMutableArray alloc] init];
        _animationDuration = 0.3;
        _contentSubViewWidthConstraints = [[NSMutableArray alloc] init];
        self.delegate = self;
        [self setPagingEnabled:YES];
        self.contentView = [[UIView alloc] init];
        self.contentView.translatesAutoresizingMaskIntoConstraints = NO;
        [self addSubview:self.contentView];
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[contentView]|" options:0 metrics:nil views:@{@"contentView": self.contentView, @"mainView": self}];
        [self addConstraints:horizontalConstraints];
        self.contentViewWidthConstraint = [NSLayoutConstraint constraintWithItem:self.contentView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:0];
        [self addConstraint:self.contentViewWidthConstraint];
        NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[contentView(==mainView)]|" options:0 metrics:nil views:@{@"contentView": self.contentView, @"mainView": self}];
        [self addConstraints:verticalConstraints];
    }
    return self;
}

- (void)addContentView:(UIView *)view frame:(CGRect)frame {
    view.translatesAutoresizingMaskIntoConstraints = NO;
    [self.contentView addSubview:view];
    CGFloat width = frame.size.width;
    if (self.contentViews.count > 0) {
        UIView *previousView = self.contentViews.lastObject;
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[previousView]-0-[view]" options:0 metrics:@{@"xPos": @(self.contentViews.count * width)} views:@{@"view": view, @"previousView": previousView}];
        [self.contentView addConstraints:horizontalConstraints];
    } else {
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[view]" options:0 metrics:@{@"xPos": @(self.contentViews.count * width)} views:@{@"view": view}];
        [self.contentView addConstraints:horizontalConstraints];
    }
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:width];
    [self.contentView addConstraint:widthConstraint];
    [self.contentSubViewWidthConstraints addObject:widthConstraint];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[view]-0-|" options:0 metrics:nil views:@{@"view": view}];
    [self.contentView addConstraints:verticalConstraints];
    [self.contentViews addObject:view];
    self.contentViewWidthConstraint.constant = self.contentViews.count * self.bounds.size.width;
}

- (void)updateContentControllersFrame:(CGRect)frame {
    CGFloat width = frame.size.width;
    self.contentViewWidthConstraint.constant = self.contentViews.count * width;
    for (NSLayoutConstraint *constraint in self.contentSubViewWidthConstraints) {
        constraint.constant = width;
    }
    [self layoutIfNeeded];
    CGPoint point = self.contentOffset;
    point.x = self.pageIndex * width;
    [self setContentOffset:point animated:YES];
}

- (void)movePageToIndex:(NSInteger)index animated:(BOOL)animated {
    self.pageIndex = index;
    CGPoint point = CGPointMake((index * self.bounds.size.width), 0);
    if (animated) {
        [UIView animateWithDuration:self.animationDuration animations:^{
            self.contentOffset = point;
        }];
    } else {
        self.contentOffset = point;
    }
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    self.pageIndex = self.contentOffset.x / self.bounds.size.width;
    if (self.didSelectSegmentAtIndex) {
        self.didSelectSegmentAtIndex(nil, self.pageIndex, YES);
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:@"DidChangeSegmentIndex" object:@(self.pageIndex)];
}

@end

File: SJSegmentedScrollView/Classes/SJContentView.swift
//
//  SJContentView.swift
//  SJSegmentedScrollView
//
//  Created by Subins Jose on 10/06/16.
//  Copyright © 2016 Subins Jose. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
//    associated documentation files (the "Software"), to deal in the Software without restriction,
//    including without limitation the rights to use, copy, modify, merge, publish, distribute,
//    sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all copies or
//    substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
//  LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
//  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import UIKit

class SJContentView: UIScrollView {
    
    var pageIndex = 0
    var contentViews = [UIView]()
    var contentView: UIView!
    var contentViewWidthConstraint: NSLayoutConstraint!
    var contentSubViewWidthConstraints = [NSLayoutConstraint]()
    let animationDuration = 0.3
    var didSelectSegmentAtIndex: DidSelectSegmentAtIndex?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        delegate = self
		isPagingEnabled = true
        
        contentView = UIView()
        contentView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(contentView)
        
        let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[contentView]|",
                                                                                   options: [],
                                                                                   metrics: nil,
                                                                                   views: ["contentView": contentView, "mainView": self])
        addConstraints(horizontalConstraints)
        
        contentViewWidthConstraint = NSLayoutConstraint(item: contentView,
                                                        attribute: .width,
                                                        relatedBy: .equal,
                                                        toItem: nil,
                                                        attribute: .notAnAttribute,
                                                        multiplier: 1.0,
                                                        constant: 0)
        addConstraint(contentViewWidthConstraint)
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[contentView(==mainView)]|",
                                                                                 options: [],
                                                                                 metrics: nil,
                                                                                 views: ["contentView": contentView, "mainView": self])
        addConstraints(verticalConstraints)
    }
    
    func addContentView(_ view: UIView, frame: CGRect) {
        
        view.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(view)
        
        let width = Int(frame.size.width)
        if contentViews.count > 0 {
            
            let previousView = contentViews.last
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:[previousView]-0-[view]",
                                                                                       options: [],
                                                                                       metrics: ["xPos": (contentViews.count * width)],
                                                                                       views: ["view": view, "previousView": previousView!])
            contentView.addConstraints(horizontalConstraints)
        } else {
            
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[view]",
                                                                                       options: [],
                                                                                       metrics: ["xPos": (contentViews.count * width)],
                                                                                       views: ["view": view])
            contentView.addConstraints(horizontalConstraints)
        }
        
        let widthConstraint = NSLayoutConstraint(item: view,
                                                 attribute: .width,
                                                 relatedBy: .equal,
                                                 toItem: nil,
                                                 attribute: .notAnAttribute,
                                                 multiplier: 1.0,
                                                 constant: CGFloat(width))
        contentView.addConstraint(widthConstraint)
        contentSubViewWidthConstraints.append(widthConstraint)
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[view]-0-|",
                                                                                 options: [],
                                                                                 metrics: nil,
                                                                                 views: ["view": view])
        contentView.addConstraints(verticalConstraints)
        contentViews.append(view)
        
        contentViewWidthConstraint.constant = CGFloat(contentViews.count) * bounds.width
    }
    
    func updateContentControllersFrame(_ frame: CGRect) {
        
        let width = frame.size.width
        contentViewWidthConstraint.constant = CGFloat(contentViews.count) * width
        
        for constraint in contentSubViewWidthConstraints {
            constraint.constant = width
        }
        
        layoutIfNeeded()
        var point = contentOffset
        if UIView.userInterfaceLayoutDirection(for: self.semanticContentAttribute) == .rightToLeft {
            let newIndex = (contentViews.count - pageIndex) - 1
            point.x = CGFloat(newIndex) * width
        } else {
            point.x = CGFloat(pageIndex) * width
        }
        
        setContentOffset(point, animated: true)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func movePageToIndex(_ index: Int, animated: Bool) {
        
        pageIndex = index
        var point = CGPoint.zero
        if UIView.userInterfaceLayoutDirection(for: self.semanticContentAttribute) == .rightToLeft {
            let newIndex = (contentViews.count - index) - 1
            point = CGPoint(x: (newIndex * Int(bounds.size.width)), y: 0)
        } else {
            point = CGPoint(x: (index * Int(bounds.size.width)), y: 0)
        }
    
        if animated == true {
            UIView.animate(withDuration: animationDuration) {
                self.contentOffset = point
            }
        } else {
            contentOffset = point
        }
    }
}

extension SJContentView: UIScrollViewDelegate {
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        pageIndex = Int(contentOffset.x / bounds.size.width)
        if UIView.userInterfaceLayoutDirection(for: self.semanticContentAttribute) == .rightToLeft {
            pageIndex = (contentViews.count - 1) - pageIndex
        }
        didSelectSegmentAtIndex?(nil, pageIndex, true)
        NotificationCenter.default.post(name: Notification.Name(rawValue: "DidChangeSegmentIndex"),
                                        object: pageIndex)
    }
}
